trin-bridge/src/bridge.rs

```diff
@@ -0,0 +1,612 @@
+                if let Err(err) = self.serve_receipts(&full_headers, &mut offer_group).await {
+                    warn!("error offering receipts: {err:?}");
+                };
+                offer_group.display_stats();
+                block_index = offer_group.range.end;
+            }
+            sleep(Duration::from_secs(LATEST_BLOCK_POLL_RATE)).await;
+        }
+    }
+
+    pub async fn launch_backfill(&self, starting_epoch: Option<u64>) {
+        let latest_block = get_latest_block().await.expect(
+            "Error launching bridge in backfill mode. Unable to get latest block from provider.",
+        );
+        let mut epoch_index = match starting_epoch {
+            Some(val) => {
+                if val * EPOCH_SIZE > latest_block {
+                    panic!(
+                        "Starting epoch is greater than current epoch. 
+                        Please specify a starting epoch that begins before the current block."
+                    );
+                }
+                val
+            }
+            None => 0,
+        };
+        let current_epoch = latest_block / EPOCH_SIZE;
+        while epoch_index < current_epoch {
+            let start = epoch_index * EPOCH_SIZE;
+            let end = start + EPOCH_SIZE;
+            // Using epoch_size chunks & epoch boundaries ensures that every
+            // "chunk" shares an epoch accumulator avoiding the need to
+            // look up the epoch acc on a header by header basis
+            let blocks_to_serve = Range { start, end };
+            let full_headers = match Bridge::get_headers(&blocks_to_serve).await {
+                Ok(val) => val,
+                Err(msg) => {
+                    warn!("Error fetching headers in range: {blocks_to_serve:?} - {msg:?}. Skipping iteration.");
+                    epoch_index += 1;
+                    continue;
+                }
+            };
+            let mut offer_group = OfferGroup::new(blocks_to_serve.clone(), epoch_index);
+            if let Err(msg) = self.offer_headers(&full_headers, &mut offer_group).await {
+                warn!("Error offering headers in range: {blocks_to_serve:?} - {msg:?}. Skipping iteration.");
+                epoch_index += 1;
+                continue;
+            };
+            if let Err(err) = self.serve_bodies(&full_headers, &mut offer_group).await {
+                warn!("error offering bodies: {err:?}");
+            };
+            if let Err(err) = self.serve_receipts(&full_headers, &mut offer_group).await {
+                warn!("error offering receipts: {err:?}");
+            };
+            offer_group.display_stats();
+            epoch_index += 1;
+        }
+    }
+
+    async fn offer_headers(
+        &self,
+        full_headers: &Vec<FullHeader>,
+        offer_group: &mut OfferGroup,
+    ) -> anyhow::Result<()> {
+        info!("Offering headers in range: {:?}", offer_group.range);
+        if offer_group.range.end <= MERGE_BLOCK_NUMBER {
```

Comment:
is it possible for this range to straddle the merge boundary?  If so, does this need more nuanced logic to split the range up at the merge boundary and push part of the list through the `premerge_` and the other part through the `post_merge_` function?

Reply:
Theoretically, it shouldn't be possible.
- if `--mode=latest` - this isn't a concern since we're always operating on post-merge blocks
- if `--mode=backfill` - we offer blocks in groups called `GossipBatch` (ex `OfferGroup`). These batches are "epoch-sized" - aka they contain all of the block data within a specific epoch (8192 blocks). Once those headers/bodies/receipts are gossiped, we move onto the next batch / epoch. So, once we reach the merge boundary, the `GossipBatch` boundary will align with the merge boundary
- if `--mode=u64` - aka backfill from epoch #N. I chose to use epoch numbers here rather than starting backfill from a specific block number, so we can take advantage of the `GossipBatch` & boundary alignment described above. I can't really foresee a scenario when we need to specifically backfill from a particular block number && using its epoch number is insufficient.